sdk/offchain/src/payment_command.rs

```diff
@@ -0,0 +1,294 @@
+    // R_SOFT
+    RecieverSoft,
+    // R_SOFT_SEND
+    RecieverSoftSend,
+    // R_SEND
+    RecieverSend,
+}
+
+impl PaymentState {
+    pub fn is_valid_transition(from: Self, to: Self) -> bool {
+        use PaymentState::*;
+
+        match from {
+            SenderInit => match to {
+                RecieverAbort | RecieverSoft | RecieverSend => true,
+                _ => false,
+            },
+            SenderSoft => match to {
+                RecieverSoftSend => true,
+                _ => false,
+            },
+            SenderSoftSend => match to {
+                RecieverAbort | RecieverSend => true,
+                _ => false,
+            },
+            RecieverSoft => match to {
+                SenderSoftSend => true,
+                _ => false,
+            },
+            RecieverSoftSend => match to {
+                SenderAbort | Ready => true,
+                _ => false,
+            },
+            RecieverSend => match to {
+                SenderAbort | SenderSoft | Ready => true,
+                _ => false,
+            },
+            RecieverAbort | SenderAbort | Ready => false,
+        }
+    }
+
+    pub fn trigger_actor(&self) -> Actor {
+        match self {
+            PaymentState::SenderInit
+            | PaymentState::SenderAbort
+            | PaymentState::SenderSoft
+            | PaymentState::SenderSoftSend
+            | PaymentState::Ready => Actor::Sender,
+
+            PaymentState::RecieverAbort
+            | PaymentState::RecieverSoft
+            | PaymentState::RecieverSoftSend
+            | PaymentState::RecieverSend => Actor::Receiver,
+        }
+    }
+
+    pub fn from_payment(payment: &PaymentObject) -> Option<Self> {
+        let state = match (
+            payment.sender().status().status(),
+            payment.sender().additional_kyc_data(),
+            payment.receiver().status().status(),
+            payment.receiver().additional_kyc_data(),
+        ) {
+            // Basic Kyc exchange flow
+            (Status::NeedsKycData, None, Status::None, None) => {
+                if payment.sender().kyc_data().is_none() {
+                    return None;
+                }
+
+                PaymentState::SenderInit
+            }
+            (Status::NeedsKycData, _, Status::ReadyForSettlement, None) => {
+                if payment.receiver().kyc_data().is_none()
+                    || payment.recipient_signature().is_none()
+                {
+                    return None;
+                }
+
+                PaymentState::RecieverSend
+            }
+            // https://dip.diem.com/dip-1/#sinit---rabort-step-9 states that the abort code needs
+            // to be one of "no-kyc-needed" or "rejected" but "no-kyc-needed" isn't a valid abort
+            // code
+            (Status::NeedsKycData, _, Status::Abort, _) => PaymentState::RecieverAbort,
+            (Status::Abort, _, Status::ReadyForSettlement, _) => PaymentState::SenderAbort,
```

Comment:
I feel it should be sender abort no matter what is receiver's status. Checked python implementation, it has similar problem.